### PR TITLE
Merge partial logs for kubernetes

### PR DESCRIPTION
**Pull request by tt810**
### What does this PR do?

* Refactor on logs/decoder, logs/line_handler
* Add logic to aggregate partial logs to one full log

### Motivation

Client requests

### Additional Notes

The process logic changed to fix the issue with "...TRUNCATED..." prepend before timestamp.

### CODE REVIEW

File: pkg/logs/input/kubernetes/line_handler.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	if fullMessage == h.prefix.flag || isFullContent {
+		h.sendAndReset()
+	}
+
+	if content1 != nil {
+		h.lineBuffer.AddIncompleteLine(content1)
+		h.sendAndReset()
+	}
+}
+
```
Comment (gaetan-deputier) [inline pkg/logs/input/kubernetes/line_handler.go]: Worth sharing that method in a util file?

**Comment by gaetan-deputier** [comment]
LGTM

**Review by gaetan-deputier — APPROVED** [review]

File: pkg/logs/decoder/line_handler.go [hunk 2]
```diff
@@ -17,195 +19,203 @@ import (
 // TRUNCATED is the warning we add at the beginning or/and at the end of a truncated message
 var TRUNCATED = []byte("...TRUNCATED...")
 
-// LineHandler handles byte slices to form line output
-type LineHandler interface {
+// LineHandlerRunner handles byte slices to form line output
+type LineHandlerRunner interface {
```
Comment (ajacquemot) [inline pkg/logs/decoder/line_handler.go]: Why Runner ? Does it run something ?

**Review by ajacquemot — CHANGES_REQUESTED** [review]
Did not finish to review this, please wait before merging.

File: pkg/logs/input/kubernetes/line_handler.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+func (h *LineHandler) Start() {
+	h.nextRunner.Start()
+	go h.RunWithTimer(h.flushTimeout, h.process, h.sendAndReset)
+}
+
+// process assumes line length does not exceed contentLenLimit, since the
+// contentLenLimit normally is 256K and kubernetes log each line doesn't
+// exceed 16K.
+func (h *LineHandler) process(line []byte) {
+	content, status, timestamp, flag, err := h.ParseFull(line)
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/line_handler.go]:
```suggestion
	content, status, timestamp, tag, err := h.ParseFull(line)
```
Does kubernetes call it a `tag` or a `flag` ? Let's be consistent

File: pkg/logs/input/kubernetes/line_handler.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+// Start starts the nextRunner and start handler runner with timer.
+func (h *LineHandler) Start() {
+	h.nextRunner.Start()
+	go h.RunWithTimer(h.flushTimeout, h.process, h.sendAndReset)
+}
+
+// process assumes line length does not exceed contentLenLimit, since the
+// contentLenLimit normally is 256K and kubernetes log each line doesn't
+// exceed 16K.
+func (h *LineHandler) process(line []byte) {
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/line_handler.go]: I feel like this method is complicate to read and hard to reason about which can be caused by the design we chose, let's discuss it IRL.

**Review by ajacquemot — COMMENTED** [review]
Left some comments, I think we have all the code, let's reorganise it now to make sur it's easy to read and maintainable.

File: pkg/logs/input/kubernetes/line_handler.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+// NewLineHandler returns a initialized LineHandler with given parameters.
+func NewLineHandler(lineHandlerRunner decoder.LineHandlerRunner, flushTimeout time.Duration, contentLenLimit int) *LineHandler {
+	return &LineHandler{
+		Handler: decoder.Handler{
+			InputChan: make(chan []byte),
+		},
+		nextRunner:      lineHandlerRunner,
+		flushTimeout:    flushTimeout,
+		lineBuffer:      decoder.NewLineBuffer(),
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/line_handler.go]: do we really have to use a line buffer here ? Why does this class need to a proper go routine ?

File: pkg/logs/input/kubernetes/line_handler.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// Stop close the inout chan and stops the next runner.
+func (h *LineHandler) Stop() {
+	close(h.InputChan)
+	h.nextRunner.Stop()
+}
+
+// Start starts the nextRunner and start handler runner with timer.
+func (h *LineHandler) Start() {
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/line_handler.go]: Why does this class need a Start and Stop ?

File: pkg/logs/input/kubernetes/line_handler.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	h.nextRunner.Start()
+	go h.RunWithTimer(h.flushTimeout, h.process, h.sendAndReset)
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/line_handler.go]: Why does this class need a timer ?

**Review by ajacquemot — COMMENTED** [review]
Left a bunch of comments and a bunch of questions that I think you should address first.

File: pkg/logs/input/kubernetes/line_handler_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/decoder"
+	"github.com/stretchr/testify/assert"
+	"regexp"
+	"testing"
+	"time"
+)
+
+func TestMultiLineHandler_SendMultiKubesLines(t *testing.T) {
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/line_handler_test.go]: nits: none of our unit tests follow this format, can you tell me why we should go for this format ?  Can we stay consistent and either update all the code base (in another PR) or keep the current convention ? Thanks!

**Review by achntrl — COMMENTED** [review]
Agreed with @ajacquemot, I think we have all the code we need to make this work. 

We should setup a meeting to discuss the design of this: 

I find it weird that we pass `contentLenLimit` everywhere even though the value we use is always the same (except in tests) 

I don't like the `LineHandlerRunner` naming either, but I like the pipeline-like design of the LineHandler in k8s that forwards it to the regular LineHandler. But we should probably discuss about the packaging of this code, I'm not sure it should live in the k8s folder

**Comment by tt810** [comment]
-

**Pull request closed without merging**